GitCommands/Git/GitGpgController.cs

```diff
@@ -0,0 +1,219 @@
+
+        /// <summary>
+        /// Obtain the number of tag on current git revision.
+        /// </summary>
+        /// <returns>Returns the number of annotated or signed tag on current revision.</returns>
+        public int NumberOfTag { get; private set; }
+
+        /// <summary>
+        /// Obtain the tag verification message for all the tag in current git revision 
+        /// </summary>
+        /// <returns>Full concatenated string coming from GPG analysis on all tags on current git revision.</returns>
+        public string TagVerifyMessage { get; private set; }
+
+        public GitGpgController(IGitModule module, GitRevision revision)
+        {
+            _module = module;
+            _revision = revision;
+
+            NumberOfTag = _revision.Refs.Count(x => x.IsTag && x.IsDereference);
+            TagVerifyMessage = "";
+        }
+
+        /// <summary>
+        /// Obtain the commit signature status on current revision.
+        /// </summary>
+        /// <returns>Enum value that indicate the gpg status for current git revision.</returns>
+        public CommitStatus CheckCommitSign()
+        {
+            CommitStatus _cmtStatus = CommitStatus.NoSignature;
+            string _gpg = _module.RunGitCmd($"log --pretty=\"format:%G?\" -1 {_revision.Guid}");
+
+            switch (_gpg)
+            {
+                case GOOD_SIGN:         // "G" for a good (valid) signature
+                    _cmtStatus = CommitStatus.GoodSignature;
+                    break;
+                case BAD_SIGN:          // "B" for a bad signature
+                case UNK_SIGN_VALIDITY: // "U" for a good signature with unknown validity 
+                case EXPIRED_SIGN:      // "X" for a good signature that has expired
+                case EXPIRED_SIGN_KEY:  // "Y" for a good signature made by an expired key
+                case REVOKED_KEY:       // "R" for a good signature made by a revoked key
+                    _cmtStatus = CommitStatus.SignatureError;
+                    break;
+                case MISSING_PUB_KEY:   // "E" if the signature cannot be checked (e.g.missing key)
+                    _cmtStatus = CommitStatus.MissingPublicKey;
+                    break;
+                case NO_SIGN:           // "N" for no signature
+                default:
+                    _cmtStatus = CommitStatus.NoSignature;
+                    break;
+            }
+
+            return _cmtStatus;
+        }
+
+
+        /// <summary>
+        /// Obtain the tag status on current revision.
+        /// </summary>
+        /// <returns>Enum value that indicate if current git revision has one tag with good signature, one tag with bad signature or more than one tag.</returns>
+        public TagStatus CheckTagSign()
+        {
+            TagStatus _tagStatus = TagStatus.OneBad;
+
+            IEnumerable<IGitRef> _usefulRef = _revision.Refs.Where(x => x.IsTag && x.IsDereference);
```

Comment:
This line duplicates the https://github.com/gitextensions/gitextensions/pull/3835/files#diff-da9708e394cb1db5b508ef9e7a8cc3a7R102 line logic. CheckTagSign assumes that NumberOfTag was checked first if greater than 0, which I find dangerous. I suggest to remove NumberOfTag at all and add a new status to the TagStatus: NoTags. This will simplify also the logic on the form where setting visibility of tagSignPicture can be done just in switch clause without the surrounding `if`

Reply:
There's a reason for this duplicated behavior.
When clicking on a commit, the revision is passed to the GitGpgController, it only counts if there are tag. 
If there are not tag, the `CheckTagSign` is not called.

Maybe can be improved having the `_usefulRef` moved into constructor and returning its count as `NumberOfTag`, what do you think ?

Reply:
I would always call CheckTagSign. It there are no tags, the CheckTagSign will return immediately returning NoTags status.

Reply:
look at newest commit